### PR TITLE
Enable keystore for autodiscover static configuration

### PR DESCRIPTION
**Pull request by ChrsMark**
Close https://github.com/elastic/beats/issues/12597.

## What this PR does
This PR makes use of keystore for containers/pods that are autodiscovered with static configurations only. Hint based configuration should not have access to the keystore for security reasons.

## How to test it manually

1. Create the keystore to store password for REDIS:
```console
./metricbeat keystore create                                                                                 
Created metricbeat keystore
./metricbeat keystore add REDIS_PASSWORD                                                                      
Enter value for REDIS_PASSWORD: 
Successfully updated the keystore
```
1. Enable `autodiscover` with static template:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            contains:
              docker.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: ["${data.host}:6379"]
              password: ${REDIS_PASSWORD}
```
2. Start metricbeat
3. Start a password redis server in a container to be autodiscovred:
```
docker run -p 6379:6379 \
--entrypoint "redis-server" redis --appendonly yes --requirepass 'passpass'
```
Check that REDIS metrics are successfully collected.
Try to start REDIS with a different password so as to make Metricbeat fail.

Now check that hints based autodiscovered containers have not access to the keystore:
1. Configure hint's based autodiscover:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      hints.enabled: true
```
2. Check that hints have no access to the keystore:
```
docker run -p 6379:6379 \
-l co.elastic.metrics/module=redis \
-l co.elastic.metrics/metricsets=info \
-l co.elastic.metrics/password='${REDIS_PASSWORD}' \
-l co.elastic.metrics/hosts='${data.host}:6379' \
--entrypoint "redis-server" redis --appendonly yes --requirepass 'passpass'
```
You should see Metribeat failing to access REDIS since the password is not retrievable.

cc: @exekias

### CODE REVIEW

File: libbeat/keystore/keystore.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ var (
 	ErrKeyDoesntExists = errors.New("cannot retrieve the key")
 )
 
+var CentralKeystore Keystore
```
Comment (exekias) [inline libbeat/keystore/keystore.go]: this is a hack. We should avoid global variables as they can cause issues when reusing code, among many other problems (I'm ok with this for testing the approach, but we need to find another solution before merging). 

I think the `Beat` object is exposing the keystore through `beat.Keystore()` method. Probably you can get it somewhere and pass it to autodiscover.

**Comment by ChrsMark** [comment]
@exekias tried to pass the `Keystore` from `instance.Beat` to `beat.Beat` and then grab it in `newMetricbeat`/`filebeat.Run` and set it to `autodiscover.Registry` 😅 . 

Not sure if this is better than the previous one but not sure if there is other obvious way so far. Feel free to comment/propose on it 🙂 .

File: filebeat/beater/filebeat.go [hunk 2]
```diff
@@ -308,6 +308,7 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 
 	var adiscover *autodiscover.Autodiscover
 	if fb.config.Autodiscover != nil {
+		autodiscover.SetAutodiscoverKeystore(b.Keystore)
```
Comment (exekias) [inline filebeat/beater/filebeat.go:311]: this could be a parameter to `NewAutodiscover` right?

File: libbeat/autodiscover/registry.go [hunk 3]
```diff
@@ -35,6 +36,8 @@ type registry struct {
 	appenders map[string]AppenderBuilder
 
 	logger *logp.Logger
+
+	keystore keystore.Keystore
```
Comment (exekias) [inline libbeat/autodiscover/registry.go:40]: you can avoid this global by passing the keystore to `NewAutodiscover`

**Comment by ChrsMark** [comment]
@exekias thanks for the suggestions! So far I changed the code so as to pass `keystore` to `NewAutodiscover` and then to each autodiscover provider (think this makes things clener and more future proof). Currently only `docker` and `kubernetes` providers utilise the keystore by adding it to each event. Happy to rethink of it either now or in a next upcoming PR I will prepare in order to introduce a new keystore backend to retrieve k8s secrets.

ps: I'm adding the docs section in a following commit.

**Comment by exekias** [comment]
nice! It would be interesting to add a test making sure that hints based autodiscover doesn't have access to the keystore

**Comment by ChrsMark** [comment]
@exekias tests and docs added, however CI is still failing at https://travis-ci.org/github/elastic/beats/jobs/680132395#L1166 but not sure if this is related to these changes.I will try to reproduce the failure locally with `master`.

**Comment by ChrsMark** [comment]
Generator tests failing on master: https://github.com/elastic/beats/pull/18014

File: libbeat/autodiscover/providers/jolokia/jolokia.go [hunk 4]
```diff
@@ -92,6 +94,7 @@ func AutodiscoverBuilder(bus bus.Bus, uuid uuid.UUID, c *common.Config) (autodis
 		builders:  builders,
 		appenders: appenders,
 		discovery: discovery,
+		keystore:  keystore,
```
Comment (exekias) [inline libbeat/autodiscover/providers/jolokia/jolokia.go:97]: you are not passing the keystore in jolokia events, should we add that too?

Comment (ChrsMark) [reply libbeat/autodiscover/providers/jolokia/jolokia.go:97]: 👍

File: libbeat/autodiscover/autodiscover.go [hunk 5]
```diff
@@ -218,7 +215,7 @@ func (a *Autodiscover) handleStart(event bus.Event) bool {
 		dynFields := a.meta.Store(hash, meta)
 
 		if a.configs[eventID][hash] != nil {
-			a.logger.Debugf("Config %v is already running", config)
+			a.logger.Debugf("Config %v is already running", common.DebugString(config, true))
```
Comment (exekias) [inline libbeat/autodiscover/autodiscover.go:218]: What would happen without this? I'm a bit concerned that at some point we add a new debug message and forget to do this

Comment (ChrsMark) [reply libbeat/autodiscover/autodiscover.go:218]: All these debug message print configuration. I was able to spot this:
`2020-04-28T11:22:21.892Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:199	Generated config: map[hosts:[172.17.0.2:6379] metricsets:[info keyspace] module:redis password:passpass]`

This one was due to this debug logging: https://github.com/elastic/beats/blob/09fd4df169324d4e02b2f3bf3c074d7501f21968/libbeat/autodiscover/autodiscover.go#L197

In this I changed all the similar ones in this method.

**Comment by ChrsMark** [comment]
jenkins, test this again please

**Comment by ChrsMark** [comment]
I will go on and merge this, since Travis succeeded except for the Generator's tests.

**Pull request merged**